google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -427,61 +434,166 @@ int computeDeadlineSeconds() {
   @InternalApi
   void extendDeadlines() {
```

Comment:
I don't have a whole lot of contex on exactly once delivery, so I might be misunderstanding this.. But, I was under the impression that AckResponses would only be sent when a user actually acks a message, or nacks a message (which is just a modack with deadline set to 0).

Here, we are internally sending ModAcks to extend deadlines of pending messages (messages that have not yet received an ack reply) for automatic lease extensions. This behavior is not typically visible to users, users just have the option to provide a maxAckExtensionPeriod to specify the maximum period of time that ack deadlines will be extended for.

When we send these modacks to the ackProcessor below, we are also sending the AckReplyConsumerWithResponse 'messageFuture' along with it. Which, I think means that we will complete the future with the incorrect response (response from internal ack extension mod ack, instead of user ack/nack).

Is there any reason that we are attaching futures to ack extension modacks at all?

Reply:
In addition to the nack case you mentioned where we would want a response sent back to the user on success/failure, we want to be able to propagate a modack failure back to the user (but not a success) so we need to pass the message future along. I can add some comments to make that more clear

Reply:
`we want to be able to propagate a modack failure back to the user (but not a success) so we need to pass the message future along.`

That's not entirely correct. We want to capture the modack errors, and if say half of the ack ids in the modack failed, we should stop retrying those half ackids and continue to extend the deadline for the other ack ids in future.

If we drop all the ack ids, this could lead to unintended redelivery.

We should NOT be propagating the errors in the internal modacks to the users.

Reply:
This was discussed offline yesterday. We want to be able to fail individual ackIds that are contained in a single modack request to the server. If the modack request fails for only some AckIds, only those will get the (correct) FAILURE message, the others will get SUCCESS.

Reply:
This was discussed offline yesterday. We want to be able to fail individual ackIds that are contained in a single modack request to the server. If the modack request fails for only some AckIds, only those will get the (correct) FAILURE message, the others will get SUCCESS.